third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs

```diff
@@ -236,6 +247,123 @@ impl<'a> FunctionGenerator<'a> {
         }
     }
 
+    /// Remove pairs of instructions that are equivalent to a no-op.
+    fn remove_redundant_instr_pairs(&mut self) {
+        let mut changed = true;
+        let mut deleted = vec![false; self.code.len()];
+        let mut no_deletions = true; // same as `deleted.iter().all(|v| !*v)`
+        let mut is_jump_target = vec![false; self.code.len()];
+        for info in self.label_info.values() {
+            if let Some(label_offs) = info.resolution {
+                if !info.references.is_empty() {
+                    is_jump_target[label_offs as usize] = true;
+                }
+            }
+        }
+        // Note: this is a worst case n^2 algorithm, but it is expected to be linear in practice.
+        // It marks instructions to be deleted lazily.
+        // If this becomes a bottleneck, we can implement a worst case linear algorithm (by
+        // maintaining a doubly-linked list of next-previous undeleted instruction indices).
+        while changed {
+            // repeat until no more deletions
+            changed = false;
+            let mut this_index = 0;
+            while this_index < (self.code.len() - 1) {
+                if deleted[this_index] {
+                    this_index += 1;
+                    continue;
+                }
+                let this_bc = &self.code[this_index];
+                if let Some(mut next_index) = deleted[(this_index + 1)..].iter().position(|v| !*v) {
+                    next_index += this_index + 1;
+                    // Note that we should not delete a sequence of pairs of instructions if we can jump to
```

Comment:
There ought to be a way to simplify this test, by carefully considering invariants, but maybe there is not.

It would be nice if any instruction you have deleted cannot be a jump target, so you can simplify this test to checking this_index and next_index.  I'd be tempted to disallow deleting jump targets until a final pass because it will simplify the code.

In any case, if (i, j) were paired and deleted, then j cannot be a jump target.  Consider code like:
```
     9:     push a
    10: L1: push x
    11:     pop x
    12: L2: push y
    13:     pop y
    14:     pop a
```
Suppose we delete (10, 11) and (12, 13), then we consider deleting (9, 14) and we need to notice that there is a deleted label so you can't delete a.  Note however, that because of the prohibition on deleting branches across labels you can do any deletions like (10, 11) and (12, 13) as a final pass; no new deletions can be enabled by them.

Not sure if that would simplify your code.

Reply:
No longer applicable due to the rewrite.